PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -204,80 +241,178 @@ public void AttachToRunspace(uint pid)
         }
 
         /// <summary>
-        /// Examines a process' modules for powershell.exe which indicates the process should
-        /// be attachable. Will generate an exception if the host process is x86 and the process
-        /// is x64.
+        /// Detaches the HostService from a local runspace
         /// </summary>
-        /// <param name="pid"></param>
-        /// <returns>True if powershell.exe is a module of the process, false otherwise.</returns>
-        public bool IsAttachable(uint pid)
+        public bool DetachFromRunspace()
         {
+            if (_callback == null)
+            {
+                _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
+            }
+
             try
             {
-                var process = Process.GetProcessById((int)pid);
-                ServiceCommon.Log("IsAttachable:" + process.ProcessName + "; id:" + process.Id);
-                foreach (ProcessModule module in process.Modules)
+                // attempt to gracefully detach the deugger
+                ClearBreakpoints();
+                ExecuteDebuggingCommand("detach", false);
+            }
+            catch (Exception)
+            {
+                // if program is running we must use stop to force the debugger to detach
+                ServiceCommon.Log("Script currently in execution, must use stop to end debugger");
+                _currentPowerShell.Stop();
+            }
+
+            using (_currentPowerShell = PowerShell.Create())
+            {
+                // scenario before exiting, used to determine if we are local detaching
+                DebugScenario preScenario = GetDebugScenario();
+
+                _attachRequestEvent.Reset();
+                _currentPowerShell.Runspace = _runspace;
+                _currentPowerShell.AddCommand("Exit-PSHostProcess");
+                _currentPowerShell.Invoke();
+
+                // wait for invoke to finish swapping the runspaces if detaching from a local process
+                if (preScenario == DebugScenario.LocalAttach)
                 {
-                    if (module.ModuleName.Equals("powershell.exe", StringComparison.Ordinal) && _installedPowerShellVersion >= RequiredPowerShellVersionForProcessAttach)
+                    _attachRequestEvent.WaitOne(DebugEngineConstants.AttachRequestEventTimeout);
+
+                    // scenario after exiting, used to determine if semaphore timed out, should not match our pre-scenario
+                    DebugScenario postScenario = GetDebugScenario();
+
+                    // make sure that the semaphore didn't just time out
+                    if (preScenario == postScenario)

```

Comment:
This brings up the question about how well we can do if that ends up timed out, I guess by just returning false won't be able to restore user back to the normal state, they probably stuck here and won't be able to do anything with the unexpected runspace anymore, it is more like an open issue, I m fine to discuss&address it later, but make sure we have an issue opened on github to track it if you feel the same way as I did.